cmd/github.go

```diff
@@ -0,0 +1,120 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/juju/ansiterm"
+	"github.com/skratchdot/open-golang/open"
+	"github.com/urfave/cli"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/github"
+
+	"github.com/manifoldco/manifold-cli/analytics"
+	"github.com/manifoldco/manifold-cli/color"
+	"github.com/manifoldco/manifold-cli/config"
+	"github.com/manifoldco/manifold-cli/session"
+	"github.com/manifoldco/manifold-cli/api"
+	"github.com/manifoldco/manifold-cli/generated/identity/client/authentication"
+	"github.com/manifoldco/manifold-cli/generated/identity/models"
+)
+
+var (
+	stateChars        = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ123456789"
+)
+
+// stateLength is the length of the generate state for GitHub
+const stateLength = 64
+const pollingTimeout = time.Minute * 2
+const pollingTick = time.Second * 5
+
+func githubWithCallback(ctx context.Context, cfg *config.Config, a *analytics.Analytics, stateType string) error {
+	// set up the oauth client
+	state := genRandomString(stateLength)
+	source := models.OAuthAuthenticationPollSourceGithub
+	_, _, pub, err := session.NewKeyMaterial(state)
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf("Could not load keys from state: %s", err), -1)
+	}
+
+	identityClient, err := api.New(api.Identity)
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf("Unable to create identity client: %s", err), -1)
+	}
+
+	uri := fmt.Sprintf("%s?cli=true&public_key=%s&type=%s&auth_token=%s", cfg.GitHubCallback, *pub, stateType, cfg.AuthToken)
+	authConfig := &oauth2.Config{
+		ClientID:    config.GitHubClientID,
+		Scopes:      []string{"user"},
+		Endpoint:    github.Endpoint,
+		RedirectURL: uri,
+	}
+
+	url := authConfig.AuthCodeURL(state, oauth2.AccessTypeOffline)
+
+	// let the user auth in the browser
+	authMsg := color.Color(ansiterm.Cyan,
+		"A browser window will open for authentication. Please close this browser when finished")
+	fmt.Println(authMsg)
+
+	time.Sleep(time.Second * 2) //
```

Comment:
Did you mean to leave that comment start?

Reply:
Comment should probably be above `time.Sleep` to tell why we're sleeping